lib/northbound_grpc.cpp

```diff
@@ -37,1315 +39,1237 @@
+		for (auto it = _cdb.begin(); it != _cdb.end(); it++)
+			delete_candidate(&it->second);
+	}
+
+	struct candidate *create_candidate(void)
+	{
+		uint64_t id = ++_next_id;
+		assert(id); // TODO: implement an algorithm for unique reusable
+			    // IDs.
+		struct candidate *c = &_cdb[id];
+		c->id = id;
+		c->config = nb_config_dup(running_config);
+		c->transaction = NULL;
+
+		return c;
+	}
+
+	void delete_candidate(struct candidate *c)
+	{
+		char errmsg[BUFSIZ] = {0};
+
+		_cdb.erase(c->id);
+		nb_config_free(c->config);
+		if (c->transaction)
+			nb_candidate_commit_abort(c->transaction, errmsg,
+						  sizeof(errmsg));
+	}
+
+	struct candidate *get_candidate(uint32_t id)
+	{
+		return _cdb.count(id) == 0 ? NULL : &_cdb[id];
+	}
+
+      private:
+	uint64_t _next_id = 0;
+	std::map<uint32_t, struct candidate> _cdb;
+};
 
-/* Thanks gooble */
 class RpcStateBase
 {
       public:
-	virtual void doCallback() = 0;
+	virtual CallState doCallback() = 0;
+	virtual void do_request(::frr::Northbound::AsyncService *service,
+				::grpc::ServerCompletionQueue *cq) = 0;
 };
 
-class NorthboundImpl;
-
-template <typename Q, typename S> class RpcState : RpcStateBase
+/*
+ * The RPC state class is used to track the execution of an RPC.
+ */
+template <typename Q, typename S> class NewRpcState : RpcStateBase
 {
+	typedef void (frr::Northbound::AsyncService::*reqfunc_t)(
+		::grpc::ServerContext *, Q *,
+		::grpc::ServerAsyncResponseWriter<S> *,
+		::grpc::CompletionQueue *, ::grpc::ServerCompletionQueue *,
+		void *);
+	typedef void (frr::Northbound::AsyncService::*reqsfunc_t)(
+		::grpc::ServerContext *, Q *, ::grpc::ServerAsyncWriter<S> *,
+		::grpc::CompletionQueue *, ::grpc::ServerCompletionQueue *,
+		void *);
+
       public:
-	RpcState(NorthboundImpl *svc,
-		 void (NorthboundImpl::*cb)(RpcState<Q, S> *))
-	    : callback(cb), responder(&ctx), async_responder(&ctx),
-	      service(svc){};
+	NewRpcState(Candidates *cdb, reqfunc_t rfunc,
+		    void (*cb)(NewRpcState<Q, S> *), const char *name)
+	    : requestf(rfunc), callback(cb), responder(&ctx),
+	      async_responder(&ctx), name(name), cdb(cdb){};
+	NewRpcState(Candidates *cdb, reqsfunc_t rfunc,
+		    void (*cb)(NewRpcState<Q, S> *), const char *name)
+	    : requestsf(rfunc), callback(cb), responder(&ctx),
+	      async_responder(&ctx), name(name), cdb(cdb){};
+
+	CallState doCallback() override
+	{
+		if (this->state == FINISH) {
+			grpc_debug("%s RPC -> [%s]", name,
+				   call_states[this->state]);
+			this->callback(this);
+			/* Is this safe? */
+			delete this;
+			return DELETED;
+		}
+
+		grpc_debug("%s RPC: %s -> PROCESS", name,
+			   call_states[this->state]);
+
+		/*
+		 * We are either in state CREATE or MORE, move back to PROCESS;
+		 * Run the callback on the main threadmaster/pthread; and
+		 * Wait for result from main thread.
+		 */
+		this->state = PROCESS;
+		thread_add_event(main_master, c_callback, (void *)this, 0,
+				 NULL);
+		pthread_mutex_lock(&this->cmux);
```

Comment:
a question about the ordering here: is it possible that the grpc thread loses, then the main pthread runs and signals the condition variable before the grpc pthread gets to the lock() call?

Reply:
I've got some changes I'm testing to clean this up.

Reply:
Ok, all cleaned up. The state transition in debug look very nice too.

Reply:
Yeah, I think some things like the MORE <-> PROCESS switcheroo may be hard to follow, but that's not your fault.